adding a beta parameter to the smooth_l1 loss fn

bdhirsh opened the pull request:
PR for issue: https://github.com/pytorch/pytorch/issues/16357

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44433 adding a beta parameter to the smooth_l1 loss fn**

Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

bdhirsh committed:
adding a beta parameter to the smooth_l1 loss fn

Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp:
```diff
@@ -46,26 +46,28 @@ static void addcdiv_cpu_kernel(TensorIterator& iter, Scalar value) {
           return norm_val * grad_output;
         else
-          return norm_val * x * grad_output;
+          return norm_val * x * grad_output / beta;
```
the smooth l1 loss when |X| < 1 now divides by the beta parameter, so the derivative does as well.

mrshenli replied:
I might be wrong. If it divides by beta in the forward pass, should we multiply the grad by beta in the backward pass?

mrshenli replied:
Chatted with @bdhirsh offline. I was wrong. Brian's current implementation is correct.

bdhirsh commented:
Making beta a double rather than a Scalar, since it's used in comparison- something that doesn't really make sense for complex types. See comment here (sorry for the duplicate PR) https://github.com/pytorch/pytorch/pull/44197#pullrequestreview-483654166

bdhirsh commented on aten/src/ATen/native/PointwiseOps.h:
```diff
@@ -11,10 +11,11 @@ struct TensorIterator;
 namespace native {
 
 using pointwise_fn = void (*)(TensorIterator&, Scalar scalar);
+using pointwise_fn_beta = void (*)(TensorIterator&, Scalar scalar, double beta);
```
lmk if you think the name should be more general, or if I just shouldn't use an alias here. Right now smooth_l1_backwards looks like the only point wise op here that takes in two scalar params

bdhirsh commented on torch/csrc/api/include/torch/nn/functional/loss.h:
```diff
@@ -316,17 +316,18 @@ inline Tensor cosine_embedding_loss(
+inline Tensor _smooth_l1_loss(const Tensor& input, const Tensor& target, double beta = 1.) {
     auto t = torch::abs(input - target);
-    return torch::where(t < 1, 0.5 * torch::pow(t, 2), t - 0.5);
+    return torch::where(t < beta, 0.5 * torch::pow(t, 2) / beta, t - 0.5 * beta);
```
it's not 100% clear to me if I need to change this, since this seems independent of the kernel code

bdhirsh commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1085,7 +1085,8 @@ Tensor l1_loss_double_backward_grad_output(const Tensor & grad, const Tensor & i
 
-Tensor smooth_l1_loss_double_backward(const Tensor & grad, const Tensor & input, const Tensor & target, int64_t reduction) {
+Tensor smooth_l1_loss_double_backward(const Tensor & grad, const Tensor & input, const Tensor & target, int64_t reduction, double beta) {
+  // TODO: figure out where to use beta
```
not 100% on how beta should be used in the double_backwards of smooth_l1_loss

mrshenli reviewed (commented):
Thanks for adding this!

We might also need to change the [SmoothL1Loss](https://github.com/pytorch/pytorch/blob/fab012aa28e58fdbf49e7da29cd7f5bff90f6bdb/torch/nn/modules/loss.py#L758) class to use this new function API, and update its doc string as well. This can come in a followup PR.

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -295,38 +295,38 @@ Tensor soft_margin_loss(
+Tensor& smooth_l1_loss_out(Tensor& result, const Tensor& input, const Tensor& target, int64_t reduction, double beta) {
   if (reduction != Reduction::None) {
-    result = at::smooth_l1_loss(input, target, reduction);
+    result = at::smooth_l1_loss(input, target, reduction, beta);
```
this looks wrong, although it was wrong before your patch.  You can't just assign the result to the C++ reference (the reference type is bogus); the semantics are to rewrite the _memory_ in `result`.

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -295,38 +295,45 @@ Tensor soft_margin_loss(
   if (reduction != Reduction::None) {
-    result = at::smooth_l1_loss(input, target, reduction);
+    Tensor loss;
+    auto iter = TensorIterator::binary_op(loss, input, target);
```
@gchanan I updated smooth_l1_loss_out to something similar to what MSE does:
- make a TensorIterator that puts the smooth l1 loss output in a temporary tensor
- use the existing `<reduction>_out` functions to put the output of the reduction in the out var.

My educated guess as to why we need the temporary variable is because `result` has a different tensor shape than the tensor output of the actually kernel function when the caller is using reductions.

gchanan commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -502,24 +502,24 @@ void minimum_kernel(TensorIterator& iter) {
-              return z < static_cast<scalar_t>(1)
-                  ? static_cast<scalar_t>(0.5) * z * z
-                  : z - static_cast<scalar_t>(0.5);
+              return z < beta
```
this looks like the math is going to be different for the vectorized vs not vectorized case.  In particular, won't the math be done in double here (if scalar_t is float), whereas it will be float in the vectorized case?

bdhirsh replied:
yes, thanks for pointing that out. I'm converting beta from double to Vec256<scalar_t> near the top.

gchanan commented on aten/src/ATen/native/cuda/BinaryMiscOpsKernels.cu:
```diff
@@ -19,11 +19,11 @@ void atan2_kernel_cuda(TensorIterator& iter) {
     gpu_kernel(iter, [] GPU_LAMBDA (scalar_t a, scalar_t b) -> scalar_t {
       auto z = ::abs(a - b);
-      return z < scalar_t(1.) ? scalar_t(0.5) * z * z : z - scalar_t(0.5);
+      return z < beta ? scalar_t(0.5) * z * z / beta : z - scalar_t(0.5) * beta;
```
similar here, isn't the math in double now?

bdhirsh replied:
In general, do we want scalar inputs to our kernels to have the same precision as the dtype of the tensor(s) they operate on? In that case it sounds a little harder since we don't want beta to take on a complex type:
- When scalar_t is a complex, use beta with double precision
- Otherwise, use beta as a scalar_t

Or would another option be just to directly make beta a float rather than a double (which I think would require reducing precision when we go from python to C++).

gchanan commented on aten/src/ATen/native/cuda/PointwiseOpsKernel.cu:
```diff
@@ -30,17 +30,17 @@ void addcdiv_cuda_kernel(TensorIterator& iter, Scalar value) {
 }
 
-void smooth_l1_backward_cuda_kernel(TensorIterator& iter, Scalar norm) {
+void smooth_l1_backward_cuda_kernel(TensorIterator& iter, Scalar norm, beta beta) {
```
this looks like a typo.  Also the same double issue.

bdhirsh replied:
I finally got cuda to build successfully this morning and point out this error to me :)

gchanan commented on torch/nn/functional.py:
```diff
@@ -2571,16 +2571,16 @@ def binary_cross_entropy_with_logits(input, target, weight=None, size_average=No
 
-def _smooth_l1_loss(input, target):
-    # type: (Tensor, Tensor) -> Tensor
+def _smooth_l1_loss(input, target, beta=1.0):
```
it doesn't look like this is actually used?

bdhirsh replied:
yes- wasn't sure originally if I could just delete it. I'll take it out now.

gchanan commented on torch/nn/functional.py:
```diff
@@ -2571,16 +2571,16 @@ def binary_cross_entropy_with_logits(input, target, weight=None, size_average=No
-def smooth_l1_loss(input, target, size_average=None, reduce=None, reduction='mean'):
-    # type: (Tensor, Tensor, Optional[bool], Optional[bool], str) -> Tensor
+def smooth_l1_loss(input, target, size_average=None, reduce=None, reduction='mean', beta=1.0):
+    # type: (Tensor, Tensor, Optional[bool], Optional[bool], str, Optional[float]) -> Tensor
```
this type doesn't look optional.

bdhirsh replied:
Fixing, thanks.

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

gchanan commented:
For posterity: https://github.com/pytorch/pytorch/issues/16357 is the issue (you should include it in the PR description).

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

bdhirsh commented:
@JackCaoG Hi Jack, I think that this PR requires a change on the XLA side- I'm updating the existing `smooth_l1_loss` operator to take in an optional `beta` param, and I noticed a failing XLA build from the PR: https://app.circleci.com/pipelines/github/pytorch/pytorch/213829/workflows/0fb2f612-971c-4fbf-aca5-974a1d85a11b/jobs/7459944.

`AtenXlaType function missed override: Tensor smooth_l1_loss(const Tensor& self, const Tensor& target, int64_t reduction);`

Let me know if I can help!

bdhirsh commented:
@robieta Posting the perf results from the torch timer after your walkthrough of how to use it. I used the torch timer to measure calls to `smooth_l1_loss` before/after the change, with 4 random tensors for 10 seconds each. and I don't think any regression here is noticeable (I noticed that the IQR's are pretty low for a given run, but the medians tended to move around a little more when I tried running each test multiple times). We'd expect maybe a small increase since the `smooth_l1_loss` call now requires an extra multiply/divide, but nothing here looks like a big change. Let me know if anything looks suspicious/wrong to you!

(I think that the warnings in the screenshot below are fine, since we only really care about the performance and not correctness of the op call).

Before:
(cpu)
<img width="595" alt="image" src="https://user-images.githubusercontent.com/16311747/93255204-8eefe500-f767-11ea-88a4-fb0ae461d484.png">

(cuda)
<img width="595" alt="image" src="https://user-images.githubusercontent.com/16311747/93238925-17628b80-f750-11ea-8a82-4f6ec24c41dc.png">


After:
(cpu)
<img width="588" alt="image" src="https://user-images.githubusercontent.com/16311747/93278531-f4a59680-f792-11ea-96e1-a638aa3f5e3e.png">

(cuda)
<img width="633" alt="image" src="https://user-images.githubusercontent.com/16311747/93278553-08e99380-f793-11ea-87bd-a84f16c6dd0f.png">

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"

PR for issue: https://github.com/pytorch/pytorch/issues/16357


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

ezyang commented:
Yes perf looks within noise. LGTM!

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"

PR for issue: https://github.com/pytorch/pytorch/issues/16357


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

JackCaoG commented:
> @JackCaoG Hi Jack, I think that this PR requires a change on the XLA side- I'm updating the existing `smooth_l1_loss` operator to take in an optional `beta` param, and I noticed a failing XLA build from the PR: https://app.circleci.com/pipelines/github/pytorch/pytorch/213829/workflows/0fb2f612-971c-4fbf-aca5-974a1d85a11b/jobs/7459944.
> 
> `AtenXlaType function missed override: Tensor smooth_l1_loss(const Tensor& self, const Tensor& target, int64_t reduction);`
> 
> Let me know if I can help!

Thanks for the heads-up, I will make the corresponding change on the XLA side.

mrshenli commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -914,20 +914,20 @@ Tensor l1_loss_double_backward_grad_output(const Tensor & grad, const Tensor & i
+Tensor smooth_l1_loss_double_backward(const Tensor & grad, const Tensor & input, const Tensor & target, int64_t reduction, double beta) {
   auto d = (input - target).abs();
-  auto grad_input = grad * (d < 1).type_as(grad);
+  auto grad_input = grad * (d < beta).type_as(grad) / beta;
```
can beta be 0 here?

bdhirsh replied:
yep, good catch. I'll try rewriting it in a way to prevent divide by zeros.

mrshenli commented on aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp:
```diff
@@ -46,26 +46,29 @@ static void addcdiv_cpu_kernel(TensorIterator& iter, Scalar value) {
           return norm_val * grad_output;
         else
-          return norm_val * x * grad_output;
+          return norm_val * x * grad_output / beta;
```
looks like if `x == beta == 0`, we could reach here, and then trigger division by 0?

mrshenli replied:
I might miss sth, but looks like this is different from the match given in the doc. 

```
0.5 (x_i - y_i)^2 / beta, & \text{if } |x_i - y_i| < beta 
```

Should the three condition be `x <= -beta`, `x >= beta`, and otherwise?

bdhirsh replied:
yep you're right, thanks. And using `x <= -beta` and `x >= beta` should fix the division by 0 issue.

mrshenli commented on torch/testing/_internal/common_nn.py:
```diff
@@ -3599,11 +3615,11 @@ def nll_loss_helper(input, target, weight, ignore_index):
-    output = ge_one_mask * (abs_diff - 0.5) + lt_one_mask * 0.5 * (abs_diff ** 2)
+    ge_beta_mask = (abs_diff >= beta).type_as(abs_diff)
+    lt_beta_mask = (abs_diff < beta).type_as(abs_diff)
+    output = ge_beta_mask * (abs_diff - 0.5 * beta) + lt_beta_mask * 0.5 * (abs_diff ** 2) / beta
```
Does this also need to handle divide by 0 case?

mrshenli reviewed (commented):
If the concern of division by 0 is a real, let's add a test to check where `|x - y| == 0`

mrshenli commented on torch/nn/modules/loss.py:
```diff
@@ -806,11 +810,14 @@ class SmoothL1Loss(_Loss):
     __constants__ = ['reduction']
 
-    def __init__(self, size_average=None, reduce=None, reduction: str = 'mean') -> None:
+    beta: float
```
this makes beta a class attribute instead of an instance attribute?

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"

PR for issue: https://github.com/pytorch/pytorch/issues/16357


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"

PR for issue: https://github.com/pytorch/pytorch/issues/16357


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/common_nn.py:
```diff
@@ -871,6 +871,36 @@ def smoothl1loss_no_reduce_scalar_test():
+        pickle=False)
+
+
+def smoothl1loss_zero_beta_test():
```
this needs to be added to `new_module_tests`?

mrshenli commented on torch/testing/_internal/common_nn.py:
```diff
@@ -871,6 +871,36 @@ def smoothl1loss_no_reduce_scalar_test():
+            lambda i: F.smooth_l1_loss(i, t.type_as(i), reduction='none', beta=0)),
+        cpp_function_call='''F::smooth_l1_loss(
+            i, t.to(i.options()), F::SmoothL1LossFuncOptions().reduction(torch::kNone), 0)''',
+        input_fn=lambda: torch.randn(2, 3, 4),
```
this won't trigger zero loss, right? If we want to test zero loss, will using `t.clone()` as input work?

bdhirsh replied:
Two things:

1) I started trying to modify the backwards kernels to avoid/check for divide-by-zeros, but I realized that setting beta=0 is really just the same thing as using l1_loss. I updated the definitions in `Loss.cpp` to check for that case and just call out to their corresponding `l1_loss` functions directly.

2) I added a test to the suite that uses beta=0, but I'm not testing for the case that the two tensors are the same. That's because it doesn't look like that test will play well with this test suite. The test suite asserts the logic by computing the analytical jacobian of the loss (using my backwards) as well as the numerical jacobian (using (f(x+e)-f(x-e)/(2*e))), but that logic breaks down when one of the functions isn't continuous. In this case, the second derivative of smooth_l1_loss when beta=0 isn't continuous- it's equal to infinity at x=0, and 0 everywhere else.

mrshenli commented on torch/testing/_internal/common_nn.py:
```diff
@@ -3599,11 +3630,15 @@ def nll_loss_helper(input, target, weight, ignore_index):
+    lt_beta_mask = (abs_diff < beta).type_as(abs_diff)
+    # guard against divide-by-zero when beta is zero.
+    if beta == 0:
+        output = abs_diff - 0.5 * beta
```
since beta is 0, we can remove the 0.5 * beta from the formula?

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"

PR for issue: https://github.com/pytorch/pytorch/issues/16357


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"

PR for issue: https://github.com/pytorch/pytorch/issues/16357


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding a beta parameter to the smooth_l1 loss fn"

PR for issue: https://github.com/pytorch/pytorch/issues/16357


Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

[ghstack-poisoned]

bdhirsh commented:
Hey @JackCaoG, I see that the PR on the XLA side has approval, so I'm going to start trying to merge this soon. Thanks for helping!

JackCaoG commented:
> Hey @JackCaoG, I see that the PR on the XLA side has approval, so I'm going to start trying to merge this soon. Thanks for helping!

sounds good!

mrshenli reviewed (approved):
LGTM! Let's land after the XLA PR, so that we won't break master.

JackCaoG commented:
> LGTM! Let's land after the XLA PR, so that we won't break master.

You guys can go first, in xla side I can simply merge the pr after this one merged(I think pytorch merge takes some time).

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6545,23 +6545,23 @@
   python_module: nn
 
-- func: smooth_l1_loss_backward.grad_input(Tensor grad_output, Tensor self, Tensor target, int reduction, *, Tensor(a!) grad_input) -> Tensor(a!)
+- func: smooth_l1_loss_backward.grad_input(Tensor grad_output, Tensor self, Tensor target, int reduction, float beta=1.0, *, Tensor(a!) grad_input) -> Tensor(a!)
```
nit: you shouldn't need to provide a default parameter -- we don't expose these backwards functions to python, they should only be called via autograd.  And you'll notice that reduction doesn't have a default parameter.

The parts I'm not sure about here are if TorchScript would let you serialize this function (it shouldn't) and/or if the backwards compatibility test would complain.

gchanan commented on torch/nn/modules/loss.py:
```diff
@@ -771,13 +771,15 @@ class SmoothL1Loss(_Loss):
         \begin{cases}
-        0.5 (x_i - y_i)^2, & \text{if } |x_i - y_i| < 1 \\
-        |x_i - y_i| - 0.5, & \text{otherwise }
+        0.5 (x_i - y_i)^2 / beta, & \text{if } |x_i - y_i| < beta \\
```
the above isn't accurate anymore ""Creates a criterion that uses a squared term if the absolute
    element-wise error falls below 1 and an L1 term otherwise.	    element-wise error falls below 1 and an L1 term otherwise."

gchanan replied:
we should also document the beta < 0.
